pkg/apiserver/api_installer.go

```diff
@@ -168,146 +161,76 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage RESTStorage
 
 			// list across namespace.
-			actions = appendIf(actions, action{"LIST", path, params}, storageVerbs["RESTLister"])
-			actions = appendIf(actions, action{"WATCHLIST", "/watch/" + path, params}, allowWatchList)
+			namer = scopeNaming{scope, a.group.linker, gpath.Join(a.prefix, itemPath), true}
+			actions = appendIf(actions, action{"LIST", path, params, namer}, storageVerbs["RESTLister"])
+			actions = appendIf(actions, action{"WATCHLIST", "/watch/" + path, params, namer}, allowWatchList)
+
 		} else {
 			// Handler for standard REST verbs (GET, PUT, POST and DELETE).
 			// v1beta1/v1beta2 format where namespace was a query parameter
 			namespaceParam := ws.QueryParameter(scope.ParamName(), scope.ParamDescription()).DataType("string")
 			namespaceParams := []*restful.Parameter{namespaceParam}
-			namespaceFn = func(req *restful.Request) (namespace string, err error) {
-				namespace = req.QueryParameter(scope.ParamName())
-				if len(namespace) == 0 {
-					namespace = api.NamespaceDefault
-				}
-				return
-			}
-
-			actions = appendIf(actions, action{"LIST", path, namespaceParams}, storageVerbs["RESTLister"])
-			actions = appendIf(actions, action{"POST", path, namespaceParams}, storageVerbs["RESTCreater"])
-			actions = appendIf(actions, action{"WATCHLIST", "/watch/" + path, namespaceParams}, allowWatchList)
 
 			itemPath := path + "/{name}"
 			nameParams := append(namespaceParams, nameParam)
-			nameFn = func(req *restful.Request) (namespace, name string, err error) {
-				namespace, _ = namespaceFn(req)
-				name = req.PathParameter("name")
-				if len(name) == 0 {
-					err = errEmptyName
-				}
-				return
-			}
-			generateLinkFn = func(namespace, name string) (path string, query string) {
-				path = strings.Replace(itemPath, "{name}", name, -1)
-				path = gpath.Join(a.prefix, path)
-				if len(namespace) > 0 {
-					values := make(url.Values)
-					values.Set(scope.ParamName(), namespace)
-					query = values.Encode()
-				}
-				return
-			}
-
-			actions = appendIf(actions, action{"GET", itemPath, nameParams}, storageVerbs["RESTGetter"])
-			actions = appendIf(actions, action{"PUT", itemPath, nameParams}, storageVerbs["RESTUpdater"])
-			actions = appendIf(actions, action{"DELETE", itemPath, nameParams}, storageVerbs["RESTDeleter"])
-			actions = appendIf(actions, action{"WATCH", "/watch/" + itemPath, nameParams}, storageVerbs["ResourceWatcher"])
-			actions = appendIf(actions, action{"REDIRECT", "/redirect/" + itemPath, nameParams}, storageVerbs["Redirector"])
-			actions = appendIf(actions, action{"PROXY", "/proxy/" + itemPath + "/{path:*}", nameParams}, storageVerbs["Redirector"])
-			actions = appendIf(actions, action{"PROXY", "/proxy/" + itemPath, nameParams}, storageVerbs["Redirector"])
+			namer := legacyScopeNaming{scope, a.group.linker, gpath.Join(a.prefix, itemPath)}
+
+			actions = appendIf(actions, action{"LIST", path, namespaceParams, namer}, storageVerbs["RESTLister"])
```

Comment:
Doesn't have to be done for this PR, but it would be useful to factor out the common lines of code between the two cases. Or maybe we should just eliminate v1beta1. :-)